sql: support partitioned hash sharded index

chengxiong-ruan opened the pull request:
Release note (sql change): Previously, crdb blocked users from creating
hash sharded index in all kinds of partitioned tables including implict
partitioned tables using `PARTITION ALL BY` or `REGIONAL BY ROW`. Now
we turn on the support of hash sharded index in implicit partitioned
tables. Which means primary key cannot be hash sharded if a table is
explicitly partitioned with `PARTITION BY` or an index cannot be hash
sharded if the index is explicitly partitioned with `PARTITION BY`.
Paritioning columns cannot be placed explicitly as key columns of a
hash sharded index as well, including regional-by-row table's `crdb_region`
column. When a hash sharded index is partitioned, ranges are pre-split 
within every single possible partition on shard boundaries. Each partition
is split up to 16 ranges, otherwise split into the number bucket count ranges.

chengxiong-ruan committed:
sql: support partitioned hash sharded index

Validations are added based on 2 rules:
1. hash sharded index cannot work with any explicit partitioning
2. hash sharded index cannot have partitioning field as explicit key column

Also have a minor fix on how we calculate number of implicit columns.

Release note (sql change): Previously, crdb blocked users from creating
hash sharded index in all kinds of partitioned tables including implict
partitioned tables using `PARTITION ALL BY` or `REGIONAL BY ROW`. Now
we turn on the support of hash sharded index in implicit partitioned
tables. Which means primary key cannot be hash sharded if a table is
explicitly partitioned with `PARTITION BY` or an index cannot be hash
sharded if the index is explicitly partitioned with `PARTITION BY`.
Paritioning columns cannot be placed explicitly as key columns of a
hash sharded index as well, including regional-by-row table's `crdb_region`
column.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76358)
<!-- Reviewable:end -->

chengxiong-ruan commented:
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvf-F5UAMt4IJtfjfBD:-Mvf-F5UAMt4IJtfjfBE:b7ltwez) ([raw file](https://github.com/cockroachdb/cockroach/blob/13c2beecca2a2766e30256457b990b5a5c208f32/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L102)):*
> ```
> CREATE TABLE t_parent (
>   id INT PRIMARY KEY USING HASH,
>   part STRING CHECK (part IN ('seattle', 'new york'))
> ```

Notice that we need this check constraint to avoid a full scan when doing FK or Uniqueness check. This is only happens to `PARTITION ALL BY`.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 1295 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvsz4VP-a-LZZsJWLOz:-Mvsz4VP-a-LZZsJWLP-:b-r9dw8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a19f7b411413b5484ee643edf03ff73475260f4/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L1295)):*
> ```
> │                             estimated row count: 1 (missing stats)
> │                             table: t_unique_hash_sec_key@t_unique_hash_sec_key_pkey
> │                             spans: /"new york"/1-/"new york"/2 /"seattle"/1-/"seattle"/2
> ```

this test seems to be flaky on the spans to scan :( which I think is because optimizer think the cost is same between them? I'll probably need to comment it out.
cc @mgartner 
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 1356 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvsz8t4EZIb0nM9aoDg:-Mvsz8t4EZIb0nM9aoDh:b-cdk2ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a19f7b411413b5484ee643edf03ff73475260f4/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L1356)):*
> ```
> │                 estimated row count: 1 (missing stats)
> │                 table: t_unique_hash_sec_key@t_unique_hash_sec_key_pkey
> │                 spans: /"new york"/2-/"new york"/3 /"seattle"/2-/"seattle"/3
> ```

same flakiness
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index, line 1179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvszH4WADwhKEEk6nvJ:-MvszH4WADwhKEEk6nvK:b-cdk2ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a19f7b411413b5484ee643edf03ff73475260f4/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index#L1179)):*
> ```
> │                             estimated row count: 1 (missing stats)
> │                             table: t_unique_hash_sec_key@t_unique_hash_sec_key_pkey
> │                             spans: /"@"/1/0 /"\x80"/1/0 /"\xc0"/1/0
> ```

same flakiness
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index, line 1240 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvszIVJB5C5fQ1liuzU:-MvszIVJB5C5fQ1liuzV:b-cdk2ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a19f7b411413b5484ee643edf03ff73475260f4/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index#L1240)):*
> ```
> │                 estimated row count: 1 (missing stats)
> │                 table: t_unique_hash_sec_key@t_unique_hash_sec_key_pkey
> │                 spans: /"@"/2/0 /"\x80"/2/0 /"\xc0"/2/0
> ```

same flakiness


<!-- Sent from Reviewable.io -->

chengxiong-ruan committed:
sql: presplit implicit partitioned list hash sharded index

Release note (sql change): When a hash sharded index is partitioned,
ranges are pre-split within every single possible partition on shard
boundaries. Each partition is split up to 16 ranges, otherwise split
into the number bucket count ranges. Note that, only list partition
is being presplit, we don't presplit for range partition.

chengxiong-ruan reviewed (commented):
The 2nd commit add support of pre-splitting on shard boundaries for each possible partitioning prefix.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-:-MvvTydr3ovkndFargO_:b-sw9rzm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @mgartner)



<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
This will need another, more in-depth review but I gave it a look because I was curious.

> Reviewed 13 of 21 files at r1, 3 of 7 files at r2, 1 of 1 files at r3, 5 of 5 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-:-MvxCX6j3vnhRbXrT9Qo:b-g8zimo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @mgartner)

___
*[-- commits, line 25 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxCPeB0VpYT35_2LlB):*
Your second commit is missing a title.
___
*[pkg/sql/create_index.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx8B6qCPcah2jRxKvK:-Mvx8B6qCPcah2jRxKvL:b-txc6ld) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_index.go#L521)):*
> ```Go
> 			return nil, nil, pgerror.New(
> 				pgcode.FeatureNotSupported,
> 				"partitioning field cannot be hash sharded index key column",
> ```

I suggest you reword this because it might be confusing. What you're trying to convey is that you cannot create this sharded index on these columns unless none of them are featured in the PARTITION ALL BY clause.
___
*[pkg/sql/create_table.go, line 1255 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx9tT97XFFLKgkzd9o:-Mvx9tT97XFFLKgkzd9p:b-kglc4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1255)):*
> ```Go
> // A non-nil regionConfig represents the region configuration of a multi-region
> // db. A nil regionConfig means current db is not multi-regional.
> //
> ```

I appreciate this comment. Doesn't `n.Locality != nil` imply that `regionConfig` is also not nil, though? This isn't your doing and I may be wrong so feel free to ignore this comment.
___
*[pkg/sql/create_table.go, line 1677 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxAGOj-T-7tUqub6d4:-MvxAGOkCZWLJ685sNC3:b-6tp3se) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1677)):*
> ```Go
> 			return nil, pgerror.New(
> 				pgcode.FeatureNotSupported,
> 				"partitioning field cannot be hash sharded index key column",
> ```

Same comment as previously.
___
*[pkg/sql/schema_changer.go, line 2681 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxC-_vCi8Pqnk4KFST:-MvxC-_vCi8Pqnk4KFSU:bnjx20p) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d4473694c379ecce60a52bda47b9b0fa934045/pkg/sql/schema_changer.go#L2681)):*
> ```Go
> 							return nil
> 						},
> 					)
> ```

Shouldn't there be a `ForEachRange` loop also? Or is that simply not relevant here because sharding & partitioning can only simultaneously be a thing for REGIONAL BY ROW and hence list-partitioning?
___
*[pkg/sql/catalog/descpb/structured.proto, line 1172 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxBQPUAzxnV5LwIdSq:-MvxBQPUAzxnV5LwIdSr:b-qacdbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/catalog/descpb/structured.proto#L1172)):*
> ```protobuf
>   // set on the table. This means that all indexes implicitly inherit all
>   // partitioning from the PARTITION ALL BY clause or region configuration.
>   optional bool partition_all_by = 44 [(gogoproto.nullable)=false];
> ```

Thanks for updating this comment. This fact was hardly obvious, sadly.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index, line 1266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx7_daBTU4TPORuyB-:-Mvx7_daBTU4TPORuyB0:b-pap492) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index#L1266)):*
> ```
>                     └── • scan buffer
>                           columns: (id, email, crdb_region, crdb_internal_email_shard_16, email_new, crdb_internal_email_shard_16_cast, check1)
>                           label: buffer 1
> ```

We don't typically have a lot of EXPLAINs in logic tests, presumably because the results are so brittle. I wonder if these are just going to be a nuisance in the long run.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
Thanks for the review! Replied to some of your comments and will fix the others.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-:-MvxinPRA_OBbASIzHf9:b-ah247d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, and @postamar)

___
*[pkg/sql/create_table.go, line 1255 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx9tT97XFFLKgkzd9o:-MvxcGJqDjxWcnRkt1Vs:b-skmcz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1255)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I appreciate this comment. Doesn't `n.Locality != nil` imply that `regionConfig` is also not nil, though? This isn't your doing and I may be wrong so feel free to ignore this comment.
</blockquote></details>

Oh, `n.Locality != nil` only means user input a SQL to create a table with `LOCALITY`, but it doesn't mean the db is multi-regional. If `n.Locality != nil` and `regionConfig`, it errors out. This happens about 40 lines below, but please allow me to paste the code here:
```go
if n.Locality != nil && regionConfig == nil &&
  !opts.bypassLocalityOnNonMultiRegionDatabaseCheck {
  return nil, errors.WithHint(pgerror.Newf(
    pgcode.InvalidTableDefinition,
    "cannot set LOCALITY on a table in a database that is not multi-region enabled",
  ),
    "database must first be multi-region enabled using ALTER DATABASE ... SET PRIMARY REGION <region>",
  )
}
```
___
*[pkg/sql/schema_changer.go, line 2681 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxC-_vCi8Pqnk4KFST:-MvxdQ5CAhxY4oP1aGfh:b-vdy12e) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d4473694c379ecce60a52bda47b9b0fa934045/pkg/sql/schema_changer.go#L2681)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't there be a `ForEachRange` loop also? Or is that simply not relevant here because sharding & partitioning can only simultaneously be a thing for REGIONAL BY ROW and hence list-partitioning?
</blockquote></details>

good callout. The reasoning here is that we'd need to sample within each range, and this implies we need a bunch of interesting logic to sampling between a pair of keys (lower bound and upper bound). We may do that but it might not be very helpful. Because the traffic tends to hit a single range still. For example, we sample 2 points `a` and `b` between `[lower_bound, upper_bound]` and we split each into ranges of 3 shards, then we'll have ranges:
```
.... some previous ranges...
[/idx_id, /idx_id/a/0)
[/idx_id/a/0, /idx_id/a/1)
[/idx_id/a/1, /idx_id/a/2)
[/idx_id/a/2, /idx_id/b/0)
[/idx_id/b/0, /idx_id/b/1)
[/idx_id/b/1, /idx_id/b/2)
..... other ranges after
```
now, say, traffic coming in inserting with some key with value between `a` and `b`, it would only hit range`[/idx_id/a/2, /idx_id/b/0)`, and the other ranges are not helpful here. We may presplit only on the sampling points instead of shard boundaries, but that's out of the scope of this pr. I'll add some comment about this, and also I forgot to add some comment we don't consider subpartition here subpartition is not supported by implicit partitioning yet.
___
*[pkg/sql/catalog/descpb/structured.proto, line 1172 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxBQPUAzxnV5LwIdSq:-MvxiGl66t4JIdde-bKO:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/catalog/descpb/structured.proto#L1172)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Thanks for updating this comment. This fact was hardly obvious, sadly.
</blockquote></details>

:)
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index, line 1266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx7_daBTU4TPORuyB-:-MvxiLBa1SmhkEJ7YZsX:b-gwy71w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index#L1266)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

We don't typically have a lot of EXPLAINs in logic tests, presumably because the results are so brittle. I wonder if these are just going to be a nuisance in the long run.
</blockquote></details>

yeah, agree! I originally thought I should put it here because I'm testing a enterprise feature. @mgartner do you want me to move these test into tests of `execbuilder`?


<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @mgartner)

___
*[pkg/sql/create_table.go, line 1255 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx9tT97XFFLKgkzd9o:-MvxpzvY-NwBk03wze_s:b-uvmqut) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1255)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Oh, `n.Locality != nil` only means user input a SQL to create a table with `LOCALITY`, but it doesn't mean the db is multi-regional. If `n.Locality != nil` and `regionConfig`, it errors out. This happens about 40 lines below, but please allow me to paste the code here:
```go
if n.Locality != nil && regionConfig == nil &&
  !opts.bypassLocalityOnNonMultiRegionDatabaseCheck {
  return nil, errors.WithHint(pgerror.Newf(
    pgcode.InvalidTableDefinition,
    "cannot set LOCALITY on a table in a database that is not multi-region enabled",
  ),
    "database must first be multi-region enabled using ALTER DATABASE ... SET PRIMARY REGION <region>",
  )
}
```
</blockquote></details>

Ah! Makes sense. Thanks.
___
*[pkg/sql/schema_changer.go, line 2681 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxC-_vCi8Pqnk4KFST:-Mvxq5076xmg7PeI6rZB:b-jxi3nw) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d4473694c379ecce60a52bda47b9b0fa934045/pkg/sql/schema_changer.go#L2681)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

good callout. The reasoning here is that we'd need to sample within each range, and this implies we need a bunch of interesting logic to sampling between a pair of keys (lower bound and upper bound). We may do that but it might not be very helpful. Because the traffic tends to hit a single range still. For example, we sample 2 points `a` and `b` between `[lower_bound, upper_bound]` and we split each into ranges of 3 shards, then we'll have ranges:
```
.... some previous ranges...
[/idx_id, /idx_id/a/0)
[/idx_id/a/0, /idx_id/a/1)
[/idx_id/a/1, /idx_id/a/2)
[/idx_id/a/2, /idx_id/b/0)
[/idx_id/b/0, /idx_id/b/1)
[/idx_id/b/1, /idx_id/b/2)
..... other ranges after
```
now, say, traffic coming in inserting with some key with value between `a` and `b`, it would only hit range`[/idx_id/a/2, /idx_id/b/0)`, and the other ranges are not helpful here. We may presplit only on the sampling points instead of shard boundaries, but that's out of the scope of this pr. I'll add some comment about this, and also I forgot to add some comment we don't consider subpartition here subpartition is not supported by implicit partitioning yet.
</blockquote></details>

Interesting! :+1: 


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @mgartner, and @postamar)

___
*[-- commits, line 25 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxCPeB0VpYT35_2LlB):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Your second commit is missing a title.
</blockquote></details>

Done.
___
*[pkg/sql/create_index.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx8B6qCPcah2jRxKvK:-MvxzQ5QAJkAq7Z7cssp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_index.go#L521)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I suggest you reword this because it might be confusing. What you're trying to convey is that you cannot create this sharded index on these columns unless none of them are featured in the PARTITION ALL BY clause.
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1255 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx9tT97XFFLKgkzd9o:-MvxzSIv36RGEYcQqexQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1255)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Ah! Makes sense. Thanks.
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1677 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxAGOj-T-7tUqub6d4:-MvxzTElBP2Yva2rdScL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1677)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Same comment as previously.
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer.go, line 2681 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxC-_vCi8Pqnk4KFST:-MvxzTt9-bBUn-a3o2L-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16d4473694c379ecce60a52bda47b9b0fa934045/pkg/sql/schema_changer.go#L2681)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Interesting! :+1: 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, and @postamar)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 1295 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvsz4VP-a-LZZsJWLOz:-Mvy0cNqDHEsSW7GBHsR:b5d8w1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a19f7b411413b5484ee643edf03ff73475260f4/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L1295)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

this test seems to be flaky on the spans to scan :( which I think is because optimizer think the cost is same between them? I'll probably need to comment it out.
cc @mgartner 
</blockquote></details>

You probably need to be explicit about the column families in the schema, because otherwise the logic tests will randomize them -- that can be a source of flakiness. The optimizer plan is deterministic.
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 859 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvy0yPr7B3t9bCfciYn:-Mvy0yPr7B3t9bCfciYo:b-fk6l6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/40faba45903fc79b747fdd5db3b5f003bb03e980/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L859)):*
> ```
>   id INT PRIMARY KEY,
>   email STRING,
>   part STRING CHECK (part IN ('seattle', 'new york'))
> ```

add the column family here and in other tables where you're showing the explain plan
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index, line 1266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx7_daBTU4TPORuyB-:-Mvy1b9zC0P_nb-9C3U9:bg37pi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index#L1266)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

yeah, agree! I originally thought I should put it here because I'm testing a enterprise feature. @mgartner do you want me to move these test into tests of `execbuilder`?
</blockquote></details>

You won't be able to move this to `execbuilder` since as you noted partitioning is an enterprise feature. But you should probably try to put the EXPLAINs in different files from the normal query results, so that we can test more configurations for the normal query results without making the EXPLAINs flaky. Notice that we have different files for this already (e.g., `regional_by_row` and `regional_by_row_query_behavior`), so you might want to follow that example.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @mgartner, @postamar, and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 1295 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvsz4VP-a-LZZsJWLOz:-Mvy_JbdDDDEinLUAgx5:b-a9cl7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a19f7b411413b5484ee643edf03ff73475260f4/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L1295)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You probably need to be explicit about the column families in the schema, because otherwise the logic tests will randomize them -- that can be a source of flakiness. The optimizer plan is deterministic.
</blockquote></details>

ah, thanks for pointing out the monster here! explicit families are added and it's not flaky anymore
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 859 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvy0yPr7B3t9bCfciYn:-Mvy_XPK8Ck6uLa90HAU:bz1lm5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/40faba45903fc79b747fdd5db3b5f003bb03e980/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L859)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

add the column family here and in other tables where you're showing the explain plan
</blockquote></details>

it worked, thanks again!
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index, line 1266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx7_daBTU4TPORuyB-:-Mvy_aenAsqPWaSzUHuD:b81puc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index#L1266)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You won't be able to move this to `execbuilder` since as you noted partitioning is an enterprise feature. But you should probably try to put the EXPLAINs in different files from the normal query results, so that we can test more configurations for the normal query results without making the EXPLAINs flaky. Notice that we have different files for this already (e.g., `regional_by_row` and `regional_by_row_query_behavior`), so you might want to follow that example.
</blockquote></details>

thanks! I moved all tests with `EXPLAIN` to separate files.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, and @postamar)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan, line 703 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw1te7zDN-f_qyILc2L:-Mw8I4WR3HWFJ_K2oDl3:b-w0gtav) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cfb251cce09a41ebe06f050a680de78f2605377/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan#L703)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

hmm... looks like CI doesn't like `auto commit` here, it failed consistently because this extra `auto commit`. But it passes at my locally consistently.
@rytaft any idea off the top your head what might cause this? Thanks!
</blockquote></details>

Does it pass both configurations for you locally? It looks like only `multiregion-9node-3region-3azs` is failing in CI. Otherwise, I don't know off the top of my head, sorry!


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
This is looking good. I still need to look at the tests in more detail, but before doing that and merging, I was hoping to have a chance to sort out the issue described here: https://github.com/cockroachdb/cockroach/pull/76697#issuecomment-1042255884 If I can't find a suitable workaround, we might have to reconsider the `UNIQUE WITHOUT INDEX` constraint method for hash-sharded indexes.

> Reviewed 12 of 21 files at r1, 2 of 7 files at r2, 3 of 5 files at r4, 2 of 4 files at r7, 1 of 5 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-:-Mw8VwDiA6it-_-T-XQJ:bg9dw92)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, and @postamar)

___
*[-- commits, line 6 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw8MQM00hoHPCqCFYTm):*
nit: "kye" => "key"
___
*[pkg/sql/opt/cat/index.go, line 163 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw8NXjFB7k8rqbgzGUQ:-Mw8NXjFB7k8rqbgzGUR:b-5j0q2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5745c464d7a4610493ea78038ac0ee58c2a6f01/pkg/sql/opt/cat/index.go#L163)):*
> ```Go
> 
> 	// ImplicitColumnCount returns the number of implicit columns at the front of
> 	// the index including implicit partitioning columns and shard column of hash
> ```

nit: "and shard column of hash sharded index" => "and shard columns of hash sharded indexes"
___
*[pkg/sql/opt/optbuilder/insert.go, line 368 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw8OfXe8B43DwUPvSQE:-Mw8OfXe8B43DwUPvSQF:bbpz1hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4132cdf88a1f8f50ebd0e734677797fc46acd7e3/pkg/sql/opt/optbuilder/insert.go#L368)):*
> ```Go
> 	// these columns will need to be fetched.
> 	primaryIndex := mb.tab.Index(cat.PrimaryIndex)
> 	if primaryIndex.ImplicitPartitioningColumnCount() > 0 {
> ```

In the case of a non-partitioned hash-sharded primary index, I don't think we need to fetch existing rows. The shard column is a function of the other PK columns, so it should be possible to perform the UPSERT fast path which blindly write the primary index entry without fetching the existing row. I don't think this is a huge concern for now, but I think it's worth leaving a todo: `TODO(mgartner): It should be possible to perform an UPSERT fast path for a non-partitioned hash-sharded primary index, but this restriction disallows it.`.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan, line 198 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw8VQqX0ZUlG01ioQln:-Mw8VQqX0ZUlG01ioQlo:bbbowdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5745c464d7a4610493ea78038ac0ee58c2a6f01/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan#L198)):*
> ```
> # using the original hash sharded index as an arbiter is not necessary.
> query T
> EXPLAIN (VERBOSE) INSERT INTO t_gen_random_uuid (val) VALUES (4321) ON CONFLICT DO NOTHING;
> ```

It would be interesting to have multi-insert-row variations for all of the `INSERT ON CONFLICT ...` and `UPSERT` tests, like we did in your previous PR. These cases are interesting because different optimizations need to be applied to make the query plan efficient.
___
*[pkg/sql/opt/exec/execbuilder/testdata/hash_sharded_index, line 263 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvspfs_3tWJ2-QcggT9:-Mw8QBC0ECG-1AedHLKz:bp87qvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/db2412ff0ae20ca5a4a5a4706fb0beaaf8545373/pkg/sql/opt/exec/execbuilder/testdata/hash_sharded_index#L263)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

looks like my change to `optIndex.ImplicitColumnCount` made the builder able to find an hash-sharded primary key arbiter would this be a concern? looks legit to me though.
</blockquote></details>

It looks semantically correct to me, but less efficient. The previous plan was a special case, the UPSERT fast path, which avoids reading from the table before performing an UPSERT. See my comment in `pkg/sql/opt/optbuilder/insert.go`. I don't think this is a huge concern, but worth leaving a TODO above this test: `TODO(mgartner): We should be able to perform the UPSERT fast path in this case, because the shard column is a function of the other PK columns.`.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, @postamar, and @rytaft)

___
*[-- commits, line 6 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw8MQM00hoHPCqCFYTm):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "kye" => "key"
</blockquote></details>

good eyes! fixed
___
*[pkg/sql/opt/optbuilder/insert.go, line 368 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw8OfXe8B43DwUPvSQE:-Mw8clDL28vK6Aj9RI3J:bborz06) ([raw file](https://github.com/cockroachdb/cockroach/blob/4132cdf88a1f8f50ebd0e734677797fc46acd7e3/pkg/sql/opt/optbuilder/insert.go#L368)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

In the case of a non-partitioned hash-sharded primary index, I don't think we need to fetch existing rows. The shard column is a function of the other PK columns, so it should be possible to perform the UPSERT fast path which blindly write the primary index entry without fetching the existing row. I don't think this is a huge concern for now, but I think it's worth leaving a todo: `TODO(mgartner): It should be possible to perform an UPSERT fast path for a non-partitioned hash-sharded primary index, but this restriction disallows it.`.
</blockquote></details>

Thanks for calling it out. TODO added
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan, line 198 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw8VQqX0ZUlG01ioQln:-Mw8ctsaBv8JIE5YCAvx:bvv6ntg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5745c464d7a4610493ea78038ac0ee58c2a6f01/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan#L198)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It would be interesting to have multi-insert-row variations for all of the `INSERT ON CONFLICT ...` and `UPSERT` tests, like we did in your previous PR. These cases are interesting because different optimizations need to be applied to make the query plan efficient.
</blockquote></details>

test cases added.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan, line 703 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mw1te7zDN-f_qyILc2L:-Mw8czRH6K7M8pLjO4wP:b-pp16ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cfb251cce09a41ebe06f050a680de78f2605377/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_hash_sharded_index_query_plan#L703)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Does it pass both configurations for you locally? It looks like only `multiregion-9node-3region-3azs` is failing in CI. Otherwise, I don't know off the top of my head, sorry!
</blockquote></details>

ha, no worries :) yeah, it passes in both configs locally
___
*[pkg/sql/opt/exec/execbuilder/testdata/hash_sharded_index, line 263 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvspfs_3tWJ2-QcggT9:-Mw8d8x6E4tOQmiuCg42:bkj76gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/db2412ff0ae20ca5a4a5a4706fb0beaaf8545373/pkg/sql/opt/exec/execbuilder/testdata/hash_sharded_index#L263)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It looks semantically correct to me, but less efficient. The previous plan was a special case, the UPSERT fast path, which avoids reading from the table before performing an UPSERT. See my comment in `pkg/sql/opt/optbuilder/insert.go`. I don't think this is a huge concern, but worth leaving a TODO above this test: `TODO(mgartner): We should be able to perform the UPSERT fast path in this case, because the shard column is a function of the other PK columns.`.
</blockquote></details>

TODO added, thanks!


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, @postamar, and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index_query_plan, line 40 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MwSCAav0sm7fPs5qyQO:-MwSCAav0sm7fPs5qyQP:bzbf3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8234562995b8a8d12467129d9704fbf0f4993084/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index_query_plan#L40)):*
> ```
> );
> 
> # TODO (mgartner): there is a lookup join with lookup condition checking every
> ```

Apologies if I added too many such `TODO`s. I wanted to add a single comment for the file, but still wanted to be more accurate and consistent with the convention on how other TODOs were added :(


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
There is still a test failed in CI due to some discrepancy on `scan` planning:
at local, with the same test configuration, I always get one `scan` scanning the 3 regions:
```
└── • scan
      columns: (email)
      estimated row count: 1 (missing stats)
      table: t_unique_hash_sec_key@idx_uniq_hash_email
      spans: /"@"/13/"some_email"/0 /"\x80"/13/"some_email"/0 /"\xc0"/13/"some_email"/0
      parallel
```
in CI it also scans the 3 regions, but it has 2 scans then make a union:
```
└── • union all
    │ columns: (email)
    │ estimated row count: 1 (missing stats)
    │ limit: 1
    │
    ├── • scan
    │     columns: (email)
    │     estimated row count: 1 (missing stats)
    │     table: t_unique_hash_sec_key@idx_uniq_hash_email
    │     spans: /"@"/13/"some_email"/0
    │
    └── • scan
          columns: (email)
          estimated row count: 1 (missing stats)
          table: t_unique_hash_sec_key@idx_uniq_hash_email
          spans: /"\x80"/13/"some_email"/0 /"\xc0"/13/"some_email"/0
          parallel
```
I've been scratching my head because could repro it :(





<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
could -> couldn't





<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Nice work! First commit <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> . I'll let someone with more experience with splitting sign off on the second commit.

I'll take a look at the flaky test and let you know what I find.

> Reviewed 2 of 5 files at r9, 15 of 15 files at r11, 3 of 3 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-:-MwSzuCz8ueVmVdy9LoK:b-toat1s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @postamar, and @rytaft)

___
*[pkg/sql/create_index.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx8B6qCPcah2jRxKvK:-MwSsS9x59KIQPbvfLNp:b6g8sbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_index.go#L521)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: Maybe we can shorten this a bit? "hash sharded indexes cannot include implicit partitioning columns from "PARTITION ALL BY" or "LOCALITY REGIONAL BY ROW"
___
*[pkg/sql/create_table.go, line 1677 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxAGOj-T-7tUqub6d4:-MwSs3rO6t6fDUiPQGFl:b-gvzeme) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1677)):*
> nit: Maybe we can shorten this a bit? "hash sharded indexes cannot include implicit partitioning columns from "PARTITION ALL BY" or "LOCALITY REGIONAL BY ROW"
___
*[pkg/sql/create_table.go, line 1894 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MwStKeqADGwhQC1Np_e:-MwStKeqADGwhQC1Np_f:b-5riwkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ab50fc898315e9ec4540fe12018a1c2ff8366ec/pkg/sql/create_table.go#L1894)):*
> ```Go
> 					return nil, pgerror.New(
> 						pgcode.FeatureNotSupported,
> 						"hash sharded indexes don't support explicit partitioning with PARTITION BY",
> ```

nit: Maybe: "hash sharded indexes cannot be explicitly partitioned"
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 134 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MwSuZMd5FgM38K6WWdq:-MwSuZMeD1-dMqJaC2da:bjraqh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ab50fc898315e9ec4540fe12018a1c2ff8366ec/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L134)):*
> ```
>   crdb_internal_c_shard_16 INT4 NOT VISIBLE NOT NULL AS (mod(fnv32(crdb_internal.datums_to_bytes(c)), 16:::INT8)) VIRTUAL,
>   CONSTRAINT t_to_be_hashed_pkey PRIMARY KEY (a ASC),
>   INDEX t_to_be_hashed_c_idx (c ASC) USING HASH WITH (bucket_count=16),
> ```

Normally there is a `CHECK` constraint created for shard columns. Why is it missing in this test and the ones below?
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index_query_plan, line 40 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MwSCAav0sm7fPs5qyQO:-MwSw7Xe0id6KTd3fudq:b7p5sb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8234562995b8a8d12467129d9704fbf0f4993084/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index_query_plan#L40)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Apologies if I added too many such `TODO`s. I wanted to add a single comment for the file, but still wanted to be more accurate and consistent with the convention on how other TODOs were added :(
</blockquote></details>

This is fine, thanks for leaving them. After this is merged I'll rebase https://github.com/cockroachdb/cockroach/pull/76817 and look into any plans that don't improve. If they don't improve, it might be related to this known issue: #63882. If so, I'll try to revive #65093.


<!-- Sent from Reviewable.io -->

rytaft commented:
For the flaky test: the plan with the UNION ALL is how we plan locality optimized search. That is the correct plan for `multiregion-9node-3region-3azs`. However, we can't yet plan locality optimized search with the `multiregion-9node-3region-3azs-tenant` configuration due to #75864, so you should remove that configuration from the top of `regional_by_row_hash_sharded_index_query_plan` (Similar to the `regional_by_row_query_behavior` file).  

Also it seems like you need these two lines to avoid flakes (also copied from `regional_by_row_query_behavior`):
```
statement ok
SET CLUSTER SETTING kv.closed_timestamp.side_transport_interval = '10ms';
SET CLUSTER SETTING kv.closed_timestamp.target_duration = '10ms';
```





<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
Thanks @rytaft, you're right! after setting those cluster settings, looks like `lookup`s and `scan`s are split into 2 ops consistently, one for local and another for remote.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-:-MwTRY0L4hDwknrjnA23:b-m1akdp)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @chengxiong-ruan, @mgartner, @postamar, and @rytaft)

___
*[pkg/sql/create_index.go, line 521 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-Mvx8B6qCPcah2jRxKvK:-MwTS0gS2a3EPaI9KH3r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_index.go#L521)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Maybe we can shorten this a bit? "hash sharded indexes cannot include implicit partitioning columns from "PARTITION ALL BY" or "LOCALITY REGIONAL BY ROW"
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1677 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MvxAGOj-T-7tUqub6d4:-MwTS1eQCEqv7Trq-7M0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbc7dd712f2a3846beeebe1248ea4b81a4512ae/pkg/sql/create_table.go#L1677)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> nit: Maybe we can shorten this a bit? "hash sharded indexes cannot include implicit partitioning columns from "PARTITION ALL BY" or "LOCALITY REGIONAL BY ROW"
</blockquote></details>

Done.
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index, line 134 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MwSuZMd5FgM38K6WWdq:-MwTManm0bd4G72ULyCA:b-lxae2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ab50fc898315e9ec4540fe12018a1c2ff8366ec/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index#L134)):*
> hash sharded indexes cannot be explicitly partitioned
yes, it's still there, we just chose not to print a hidden constraint now since `USING HASH` implies the constraint with this fix https://github.com/cockroachdb/cockroach/pull/74179
let me know if that makes sense.
___
*[pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index_query_plan, line 40 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-MwSCAav0sm7fPs5qyQO:-MwTS4n8AIx5GQfRkq6O:b-9755fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8234562995b8a8d12467129d9704fbf0f4993084/pkg/ccl/logictestccl/testdata/logic_test/partitioning_hash_sharded_index_query_plan#L40)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This is fine, thanks for leaving them. After this is merged I'll rebase https://github.com/cockroachdb/cockroach/pull/76817 and look into any plans that don't improve. If they don't improve, it might be related to this known issue: #63882. If so, I'll try to revive #65093.
</blockquote></details>

Thank you!


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r13, 3 of 3 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76358#-:-MwX4yg_CXlTOsDbN725:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, @mgartner, @postamar, and @rytaft)



<!-- Sent from Reviewable.io -->

postamar commented:
LGTM as far as I can tell.

chengxiong-ruan commented:
bors r+
